src/java/org/apache/cassandra/cql3/ColumnsExpression.java

```diff
@@ -0,0 +1,558 @@
+    {
+        // All columns must have the same type.
+        return firstColumn().kind;
+    }
+
+    /**
+     * Returns the expression kind.
+     * @return the expression kind.
+     */
+    public Kind kind()
+    {
+        return kind;
+    }
+
+    /**
+     * Returns the expression type
+     * @return the expression type
+     */
+    public AbstractType<?> type()
+    {
+        return type;
+    }
+
+    public ByteBuffer mapKey(QueryOptions options)
+    {
+        ByteBuffer key = mapKey.bindAndGet(options);
+        if (key == null)
+            throw invalidRequest("Invalid null map key for column %s", firstColumn().name.toCQLString());
+        if (key == ByteBufferUtil.UNSET_BYTE_BUFFER)
+            throw invalidRequest("Invalid unset map key for column %s", firstColumn().name.toCQLString());
+        return key;
+    }
+
+    /**
+     * Collects the column specifications for the bind variables in the map key.
+     * This is obviously a no-op if the expression is not a {@code MAP_ELEMENT} expression.
+     *
+     * @param boundNames the variables specification where to collect the
+     * bind variables of the map key in.
+     */
+    public void collectMarkerSpecification(VariableSpecifications boundNames)
+    {
+        if (mapKey != null)
+            mapKey.collectMarkerSpecification(boundNames);
+    }
+
+    /**
+     * Checks if this instance is a column level expression (single or multi-column expression).
+     * @return {@code true} if this instance is a column level expression, {@code false} otherwise.
+     */
+    public boolean isColumnLevelExpression()
+    {
+        return kind == Kind.SINGLE_COLUMN || kind == Kind.MULTI_COLUMN;
+    }
+
+    /**
+     * Adds all functions (native and user-defined) used by any component of the restriction
+     * to the specified list.
+     * @param functions the list to add to
+     */
+    public void addFunctionsTo(List<Function> functions)
+    {
+        if (mapKey != null)
+            mapKey.addFunctionsTo(functions);
+    }
+
+    /**
+     * Returns CQL representation of this expression.
+     * @return the CQL representation of this expression.
+     */
+    public String toCQLString()
+    {
+        return kind.toCQLString(columns, mapKey);
+    }
+
+    @Override
+    public String toString()
+    {
+        String prefix = kind == Kind.SINGLE_COLUMN ? "column "
+                                                   : kind == Kind.MULTI_COLUMN ? "tuple " : "";
+
+        return prefix + toCQLString();
+    }
+
+    /**
+     * Returns the column specification corresponding to this expression.
+     * @param table the table metadata
+     * @return the column specification corresponding to this expression.
+     */
+    public ColumnSpecification columnSpecification(TableMetadata table)
+    {
+        return new ColumnSpecification(table.keyspace, table.name, new ColumnIdentifier(toCQLString(), true), type) ;
+    }
+
+    public static final class Raw
```

Comment:
I would add a javadoc here explaining what `Raw` means and what it is intended for